components/usage/pkg/controller/reconciler.go

```diff
@@ -302,20 +200,20 @@ func trimStartStopTime(instances []db.WorkspaceInstance, maximumStart, minimumSt
 	return updated
 }
 
-func toSet(items []string) []string {
-	m := map[string]struct{}{}
-	for _, i := range items {
-		m[i] = struct{}{}
-	}
-
-	var result []string
-	for s := range m {
-		result = append(result, s)
-	}
-	return result
-}
-
 type TeamUsage struct {
 	TeamID           string `json:"team_id"`
 	WorkspaceSeconds int64  `json:"workspace_seconds"`
 }
+
+func groupInstancesByAttributionID(instances []db.WorkspaceInstance) map[db.AttributionID][]db.WorkspaceInstance {
+	result := map[db.AttributionID][]db.WorkspaceInstance{}
+	for _, instance := range instances {
+		if _, ok := result[instance.UsageAttributionID]; !ok {
+			result[instance.UsageAttributionID] = []db.WorkspaceInstance{}
+		}
+
+		result[instance.UsageAttributionID] = append(result[instance.UsageAttributionID], instance)
+	}
+
+	return result
+}
```

Comment:
Are you sure it makes sense performance-wise to re-implement a SQL "group by" in Go?

I wonder if maybe this should be part of the query. Relatedly, have you had a chance to chat with @geropl about the query's performance, and how to efficiently index and shard it? 👀

Reply:
You can only use group-by with summary functions (like count). That's because a group-by takes multiple rows, and turns them into a single row by aggregating them.

In practice, we need the full dataset anyway because the billing controller needs to do another pass to drop WSI which have already been billed. 

Happy to go into more details on this.